lib/src/config_concatenation.cc

```diff
@@ -0,0 +1,248 @@
+#include <hocon/config_object.hpp>
+#include <hocon/path.hpp>
+#include <internal/config_concatenation.hpp>
+#include <internal/values/config_string.hpp>
+#include <internal/default_transformer.hpp>
+#include <internal/simple_config_list.hpp>
+#include <internal/simple_config_origin.hpp>
+#include <internal/config_exception.hpp>
+#include <internal/values/config_string.hpp>
+#include <internal/resolve_result.hpp>
+#include <internal/resolve_source.hpp>
+#include <internal/resolve_context.hpp>
+
+using namespace std;
+
+namespace hocon {
+
+    config_concatenation::config_concatenation(shared_origin origin, std::vector<shared_value> pieces) :
+            config_value(move(origin)), _pieces(move(pieces)), _unmerged_values({ shared_from_this() }) {
```

Comment:
I feel like I'm overlooking a possibility here. The only use of `unmerged_values()` is in https://github.com/puppetlabs/cpp-hocon/blob/master/lib/src/values/config_value.cc#L223-L224. We could optimize for that somehow.